File: os2skoledata-azure-ad-sync/src/main/java/dk/digitalidentity/service/OS2skoledataService.java
package dk.digitalidentity.service;

import com.microsoft.graph.models.User;
import dk.digitalidentity.config.OS2skoledataAzureADConfiguration;
import dk.digitalidentity.service.model.DBGroup;
import dk.digitalidentity.service.model.Institution;
import dk.digitalidentity.service.model.DBUser;
import dk.digitalidentity.service.model.ModificationHistory;
import dk.digitalidentity.service.model.enums.Action;
import dk.digitalidentity.service.model.enums.EntityType;
import dk.digitalidentity.service.model.enums.Role;
import dk.digitalidentity.service.model.enums.SetFieldType;
import dk.digitalidentity.service.model.enums.StudentRole;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Slf4j
@Service
public class OS2skoledataService {

	@Autowired
	private OS2skoledataAzureADConfiguration config;

	public List<Institution> getInstitutions() throws Exception {
		log.info("Fetching institutions");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/institutions";

		ResponseEntity<Institution[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, Institution[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch institutions. Will not update");
		}

		log.info("Finished fetching institutions");
		return Arrays.asList(response.getBody());
	}

	public List<DBGroup> getClassesForInstitution(Institution institution) throws Exception {
		log.info("Fetching classes for institutions " + institution.getInstitutionName());
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/groups?institutionNumber=" + institution.getInstitutionNumber();

		ResponseEntity<DBGroup[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, DBGroup[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch groups for institution " + institution.getInstitutionName() + ". Will not update");
		}

		log.info("Finished fetching classes for institutions " + institution.getInstitutionName());
		return Arrays.asList(response.getBody());
	}

	public List<DBUser> getUsersForInstitution(Institution institution) throws Exception {

		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/persons?institutionNumber=" + institution.getInstitutionNumber();

		ResponseEntity<DBUser[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, DBUser[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch institutions. Will not update");
		}

		log.info("Finished fetching users for institutions " + institution.getInstitutionName());
		return Arrays.asList(response.getBody());
	}

	record UsernameRequest(long personDatabaseId, String username) {}
	public void setUsernameOnUser(long databaseId, String username) throws Exception {
		if (config.getAzureAd().isUserDryRun()) {
			log.info("UserDryRun: Would set username " + username + " on user with database id " + databaseId + " in OS2skoledata.");
			return;
		}
		UsernameRequest usernameRequest = new UsernameRequest(databaseId, username);
		HttpEntity<UsernameRequest> request = new HttpEntity<>(usernameRequest, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/person/username";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set username on user in OS2skoledata. Message: " + response.getBody());
		}
	}

	public long getHead() throws Exception {
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/head";

		ResponseEntity<Long> response = new RestTemplate().exchange(query, HttpMethod.GET, request, Long.class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch head. Will not update");
		}

		return response.getBody();
	}

	record HeadRequest(long head) {}
	public void setHead(long head, String institutionNumber) throws Exception {
		if (config.getAzureAd().isUserDryRun()) {
			log.info("UserDryRun: Would set head " + head + " on institution " + institutionNumber);
			return;
		}
		HeadRequest headRequest = new HeadRequest(head);
		HttpEntity<HeadRequest> request = new HttpEntity<>(headRequest, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/head/" + institutionNumber;

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set head in OS2skoledata for institution: " + institutionNumber + ". Message: " + response.getBody());
		}
	}

	record SetFieldRequest(long id, EntityType entityType, SetFieldType fieldType, String value) {}
	public void setFields(long id, EntityType entityType, SetFieldType fieldType, String value) throws Exception {
		SetFieldRequest setFieldRequest = new SetFieldRequest(id, entityType, fieldType, value);
		HttpEntity<SetFieldRequest> request = new HttpEntity<>(setFieldRequest, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/setfield";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set field on in OS2skoledata. Message: " + response.getBody());
		}
	}

	public List<ModificationHistory> getChanges(String institutionNumber) throws Exception {
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/changes/" + institutionNumber;

		ResponseEntity<ModificationHistory[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, ModificationHistory[].class);
		if (response.getStatusCode().equals(HttpStatus.OK) && response.getBody() != null) {
			return Arrays.asList(response.getBody());
		} else {
			if (response.getStatusCode().equals(HttpStatus.CONFLICT)) {
				throw new Exception("Do a full sync");
			} else {
				throw new Exception("Failed to fetch changes. Will not delta update");
			}
		}
	}

	record ChangeRequest(List<Long> ids){}
	public List<DBUser> getChangedUsers(List<Long> changedPersonIds) throws Exception {
		ChangeRequest error = new ChangeRequest(changedPersonIds);
		HttpEntity<ChangeRequest> request = new HttpEntity<>(error, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/changes/persons";

		ResponseEntity<DBUser[]> response = new RestTemplate().exchange(query, HttpMethod.POST, request, DBUser[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch changed users. Will not delta update");
		}
		return Arrays.asList(response.getBody());
	}

	public List<Institution> getChangedInstitutions(List<Long> changedInstitutionIds) throws Exception {
		ChangeRequest error = new ChangeRequest(changedInstitutionIds);
		HttpEntity<ChangeRequest> request = new HttpEntity<>(error, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/changes/persons";

		ResponseEntity<Institution[]> response = new RestTemplate().exchange(query, HttpMethod.POST, request, Institution[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch changed institutions. Will not delta update");
		}
		return Arrays.asList(response.getBody());
	}

	public List<DBGroup> getChangedGroups(List<Long> changedGroupIds) throws Exception {
		ChangeRequest error = new ChangeRequest(changedGroupIds);
		HttpEntity<ChangeRequest> request = new HttpEntity<>(error, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/changes/persons";

		ResponseEntity<DBGroup[]> response = new RestTemplate().exchange(query, HttpMethod.POST, request, DBGroup[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch changed groups. Will not delta update");
		}
		return Arrays.asList(response.getBody());
	}

	public List<String> getLockedUsernames() throws Exception {
		log.info("Fetching locked usernames");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/locked/usernames";

		ResponseEntity<String[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch locked usernames. Will not update");
		}

		log.info("Finished fetching locked usernames");
		return Arrays.asList(response.getBody());
	}

	public List<String> getAllUsernames() throws Exception {
		log.info("Fetching all usernames");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/usernames/all";

		ResponseEntity<String[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch all usernames. Will not update");
		}

		log.info("Finished fetching all usernames");
		return Arrays.asList(response.getBody());
	}

	public List<String> getLockedGroupIds() throws Exception {
		log.info("Fetching locked group ids");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/locked/groups/azure";

		ResponseEntity<String[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch locked group ids. Will not update");
		}

		log.info("Finished fetching locked group ids");
		return Arrays.asList(response.getBody());
	}

	public List<String> getLockedTeamIds() throws Exception {
		log.info("Fetching locked team ids");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/locked/teams/azure";

		ResponseEntity<String[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch locked team ids. Will not update");
		}

		log.info("Finished fetching locked team ids");
		return Arrays.asList(response.getBody());
	}

	record ActionRequest(String username, Action action, String integrationType) {}
	public void setActionOnUser(String username, Action action) throws Exception {
		if (config.getAzureAd().isUserDryRun()) {
			log.info("UserDryRun: Would set action " + action + " on user with " + username + " in OS2skoledata.");
			return;
		}
		ActionRequest actionRequest = new ActionRequest(username, action, "AZURE");
		HttpEntity<ActionRequest> request = new HttpEntity<>(actionRequest, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/person/action";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set action on user in OS2skoledata. Message: " + response.getBody());
		}
	}

	public List<String> getKeepAliveUsernames() throws Exception {
		log.info("Fetching keep alive usernames");
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/keepalive";

		ResponseEntity<String[]> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String[].class);
		if (!response.getStatusCode().equals(HttpStatus.OK) || response.getBody() == null) {
			throw new Exception("Failed to fetch keep alive usernames. Will not update");
		}

		log.info("Finished fetching keep alive usernames");
		return Arrays.asList(response.getBody());
	}

	record SetGroupIdentifierRequest(long institutionId, String groupKey, String groupEmail) {}
	public void setGroupIdentifier(long institutionId, String groupKey, String value) throws Exception {
		SetGroupIdentifierRequest setGroupIdentifierRequest = new SetGroupIdentifierRequest(institutionId, groupKey, value);
		HttpEntity<SetGroupIdentifierRequest> request = new HttpEntity<>(setGroupIdentifierRequest, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/setgroupemail?integration=AZURE";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set identifier on group in OS2skoledata. Message: " + response.getBody());
		}
	}

	public String getSchemaId() throws Exception {
		HttpEntity<String> request = new HttpEntity<>(getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/settings/azure/schemaId";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.GET, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to fetch schema id");
		}

		return response.getBody();
	}

	public void setSchemaId(String schemaId) throws Exception {
		HttpEntity<String> request = new HttpEntity<>(schemaId, getHeaders());
		String query = config.getOs2skoledata().getBaseUrl() + "/api/settings/azure/schemaId";

		ResponseEntity<String> response = new RestTemplate().exchange(query, HttpMethod.POST, request, String.class);
		if (!response.getStatusCode().equals(HttpStatus.OK)) {
			throw new Exception("Failed to set schemaID in OS2skoledata. Message: " + response.getBody());
		}
	}

	private HttpHeaders getHeaders() {
		HttpHeaders headers = new HttpHeaders();
		headers.add("apiKey", config.getOs2skoledata().getApiKey());
		headers.add("Content-Type", "application/json");
		headers.add("Accept", "application/json");
		return headers;
	}
}

File: os2skoledata-google-workspace-sync/Services/OS2skoledata/OS2skoledataService.cs
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using os2skoledata_google_workspace_sync.Services.OS2skoledata.Model;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;

namespace os2skoledata_google_workspace_sync.Services.OS2skoledata
{
    internal class OS2skoledataService : ServiceBase<OS2skoledataService>
    {
        private readonly Uri baseUri;
        private readonly string apiKey;
        private readonly bool userDryRun;
        private readonly JsonSerializerSettings jsonSerializerSettings = new JsonSerializerSettings() { NullValueHandling = NullValueHandling.Ignore };

        public OS2skoledataService(IServiceProvider sp) : base(sp)
        {
            baseUri = new Uri(settings.OS2skoledataSettings.BaseUrl);
            apiKey = settings.OS2skoledataSettings.ApiKey;
            userDryRun = settings.WorkspaceSettings.UserDryRun;
        }

        public List<Institution> GetInstitutions()
        {
            logger.LogInformation("Fetching institutions from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/institutions"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var institutionsArray = JsonConvert.DeserializeObject<Institution[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching institutions from OS2skoledata");
            return new List<Institution>(institutionsArray);
        }

        public List<DBGroup> GetClassesForInstitution(Institution institution)
        {
            logger.LogInformation($"Fetching groups from {institution.InstitutionName} in OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/groups?institutionNumber={institution.InstitutionNumber}"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var groupArray = JsonConvert.DeserializeObject<DBGroup[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished fetching groups from {institution.InstitutionName} in OS2skoledata");
            return new List<DBGroup>(groupArray);
        }

        public List<DBUser> GetUsersForInstitution(Institution institution)
        {
            logger.LogInformation($"Fetching people from {institution.InstitutionName} in OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/persons?institutionNumber={institution.InstitutionNumber}"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var userArray = JsonConvert.DeserializeObject<DBUser[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished fetching people from {institution.InstitutionName} in OS2skoledata");
            return new List<DBUser>(userArray);
        }

        public void SetFields(long id, EntityType type, SetFieldType fieldType, string value)
        {
            SetFieldRequest request = new SetFieldRequest();
            request.Id = id;
            request.EntityType = type.ToString();
            request.FieldType = fieldType.ToString();
            request.Value = value;


            string json = JsonConvert.SerializeObject(request, getSerializerSettings());
            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/setfield"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public void SetGroupEmail(long institutionId, string groupKey, string email)
        {
            SetGroupEmailRequest request = new SetGroupEmailRequest();
            request.InstitutionId = institutionId;
            request.GroupKey = groupKey;
            request.GroupEmail = email;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/setgroupemail?integration=GW"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public void ReportError(string errorMsg)
        {
            ErrorRequest request = new ErrorRequest();
            request.Message = errorMsg;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/reporterror"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public long GetHead()
        {
            logger.LogInformation("Fetching head from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/head"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var head = JsonConvert.DeserializeObject<long>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching head from OS2skoledata");
            return head;
        }

        public void SetHead(long head, string institutionNumber)
        {
            if (userDryRun)
            {
                logger.LogInformation($"DryRun: would have sat head {head} for institution with number {institutionNumber} in OS2skoledata");
            }
            else
            {
                HeadRequest request = new HeadRequest();
                request.Head = head;

                using var httpClient = GetHttpClient();
                var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
                var response = httpClient.PostAsync(new Uri(baseUri + "api/head/" + institutionNumber), content);
                response.Wait();
                response.Result.EnsureSuccessStatusCode();
            }
        }

        public List<ModificationHistory> GetChangesForInstitution(string institutionNumber)
        {
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/changes/" + institutionNumber));
            response.Wait();
            if (response.Result.IsSuccessStatusCode)
            {
                var responseString = response.Result.Content.ReadAsStringAsync();
                responseString.Wait();
                var changeArray = JsonConvert.DeserializeObject<ModificationHistory[]>(responseString.Result, jsonSerializerSettings);

                return new List<ModificationHistory>(changeArray);
            }
            else
            {
                if (response.Result.StatusCode.Equals(HttpStatusCode.Conflict))
                {
                    throw new Exception("Do a full sync");
                }
                else
                {
                    throw new Exception("Failed to fetch changes from os2skoledata. Http: " + response.Result.StatusCode);
                }
            }

        }

        public List<Institution> GetChangedInstitutions(List<long> ids)
        {
            ChangeRequest request = new ChangeRequest();
            request.Ids = ids;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri, $"api/changes/institutions"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();
            var changeArray = JsonConvert.DeserializeObject<Institution[]>(responseString.Result, jsonSerializerSettings);

            return new List<Institution>(changeArray);
        }

        public List<DBUser> GetChangedUsers(List<long> ids)
        {
            ChangeRequest request = new ChangeRequest();
            request.Ids = ids;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri, $"api/changes/persons"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();
            var changeArray = JsonConvert.DeserializeObject<DBUser[]>(responseString.Result, jsonSerializerSettings);

            return new List<DBUser>(changeArray);
        }

        public List<DBGroup> GetChangedGroups(List<long> ids)
        {
            ChangeRequest request = new ChangeRequest();
            request.Ids = ids;


            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri, $"api/changes/groups"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();
            var changeArray = JsonConvert.DeserializeObject<DBGroup[]>(responseString.Result, jsonSerializerSettings);

            return new List<DBGroup>(changeArray);
        }

        public List<string> GetLockedUsernames()
        {
            logger.LogInformation($"Fetching locked usernames from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/locked/usernames"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var stringArray = JsonConvert.DeserializeObject<string[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished locked usernames from OS2skoledata");
            return new List<string>(stringArray);
        }

        public List<string> GetLockedGroupEmails()
        {
            logger.LogInformation($"Fetching locked group emails from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/locked/groups/workspace"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var stringArray = JsonConvert.DeserializeObject<string[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished locked group emails from OS2skoledata");
            return new List<string>(stringArray);
        }

        public List<string> GetLockedDriveIds()
        {
            logger.LogInformation($"Fetching locked drive ids from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/locked/drives/workspace"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var stringArray = JsonConvert.DeserializeObject<string[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished locked drive ids from OS2skoledata");
            return new List<string>(stringArray);
        }

        public void SetUsernameOnUser(string username, long id)
        {
            UsernameRequest request = new UsernameRequest();
            request.PersonDatabaseId = id;
            request.Username = username;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/person/username"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public List<string> GetAllUsernames()
        {
            logger.LogInformation($"Fetching all usernames from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/usernames/all"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var usernameArray = JsonConvert.DeserializeObject<string[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"finished fetching all usernames OS2skoledata");
            return new List<string>(usernameArray);
        }

        public void SetActionOnUser(string username, ActionType action)
        {
            ActionRequest request = new ActionRequest();
            request.Action = action;
            request.Username = username;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/person/action"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public List<string> GetKeepAliveUsernames()
        {
            logger.LogInformation($"Fetching keep alive usernames from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, $"api/keepalive"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var usernameArray = JsonConvert.DeserializeObject<string[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation($"Finshed fetching keep alive usernames OS2skoledata");
            return new List<string>(usernameArray);
        }

        public List<ClassroomPendingChange> GetPendingClassroomChanges()
        {
            logger.LogInformation($"Fetching all pending classroom changes from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/classrooms/pending"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var changesArray = JsonConvert.DeserializeObject<ClassroomPendingChange[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching all pending classroom changes from OS2skoledata");
            return new List<ClassroomPendingChange>(changesArray);
        }

        public void SetClassroomActionStatus(long id, ClassroomActionStatus status, string error)
        {
            SetClassroomStatusRequest request = new SetClassroomStatusRequest();
            request.Id = id;
            request.Status = status;
            request.ErrorMessage = error;

            using var httpClient = GetHttpClient();
            var content = new StringContent(JsonConvert.SerializeObject(request, getSerializerSettings()), Encoding.UTF8, "application/json");
            var response = httpClient.PostAsync(new Uri(baseUri + "api/classrooms/status"), content);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public bool GetPerformYearChange()
        {
            logger.LogInformation("Fetching perform year change from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/yearchange"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var perform = JsonConvert.DeserializeObject<bool>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching perform year change from OS2skoledata");
            return perform;
        }

        public void SetPerformedYearChange()
        {
            using var httpClient = GetHttpClient();
            var response = httpClient.PostAsync(new Uri(baseUri + "api/yearchange?deletedfoldersandgroups=true"), null);
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
        }

        public List<FolderOrGroupDTO> GetYearChangeGroupsAndFoldersForDeletion()
        {
            logger.LogInformation($"Fetching all groups and folders for deletion from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/yearchange/foldersandgroups/delete"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var changesArray = JsonConvert.DeserializeObject<FolderOrGroupDTO[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching all groups and folders for deletion from OS2skoledata");
            return new List<FolderOrGroupDTO>(changesArray);
        }

        public List<FolderOrGroupDTO> GetAllYearlyClassFolders()
        {
            logger.LogInformation($"Fetching all yearly class folders from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/yearchange/foldersandgroups/folders"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var changesArray = JsonConvert.DeserializeObject<FolderOrGroupDTO[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching all yearly class folders from OS2skoledata");
            return new List<FolderOrGroupDTO>(changesArray);
        }

        public List<FolderOrGroupDTO> GetAllYearlyClassGroups()
        {
            logger.LogInformation($"Fetching all yearly class groups from OS2skoledata");
            using var httpClient = GetHttpClient();
            var response = httpClient.GetAsync(new Uri(baseUri, "api/yearchange/foldersandgroups/groups"));
            response.Wait();
            response.Result.EnsureSuccessStatusCode();
            var responseString = response.Result.Content.ReadAsStringAsync();
            responseString.Wait();

            var changesArray = JsonConvert.DeserializeObject<FolderOrGroupDTO[]>(responseString.Result, jsonSerializerSettings);

            logger.LogInformation("finished fetching all yearly class groups from OS2skoledata");
            return new List<FolderOrGroupDTO>(changesArray);
        }

        private HttpClient GetHttpClient()
        {
            var handler = new HttpClientHandler();
            handler.ClientCertificateOptions = ClientCertificateOption.Manual;
            handler.ServerCertificateCustomValidationCallback =
                (httpRequestMessage, cert, cetChain, policyErrors) =>
                {
                    return true;
                };

            var httpClient = new HttpClient(handler);
            httpClient.DefaultRequestHeaders.Add("ApiKey", apiKey);
            return httpClient;
        }
        private JsonSerializerSettings getSerializerSettings()
        {
            // api requires camel case
            return new JsonSerializerSettings
            {
                ContractResolver = new DefaultContractResolver
                {
                    NamingStrategy = new CamelCaseNamingStrategy()
                }
            };
        }
    }
}
